coderd/notifications/reports/generator.go

```diff
@@ -0,0 +1,296 @@
+package reports
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"io"
+	"slices"
+	"sort"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/google/uuid"
+
+	"github.com/coder/quartz"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+	"github.com/coder/coder/v2/coderd/database/dbtime"
+	"github.com/coder/coder/v2/coderd/notifications"
+	"github.com/coder/coder/v2/codersdk"
+)
+
+const (
+	delay = 15 * time.Minute
+)
+
+func NewReportGenerator(ctx context.Context, logger slog.Logger, db database.Store, enqueuer notifications.Enqueuer, clk quartz.Clock) io.Closer {
+	closed := make(chan struct{})
+
+	ctx, cancelFunc := context.WithCancel(ctx)
+	//nolint:gocritic // The system generates periodic reports without direct user input.
+	ctx = dbauthz.AsSystemRestricted(ctx)
+
+	// Start the ticker with the initial delay.
+	ticker := clk.NewTicker(delay)
+	doTick := func(start time.Time) {
+		defer ticker.Reset(delay)
```

Comment:
I don't think this is needed? The ticker ticks with the given interval without needing to reset AFAIK.

Reply:
Right, but if the routine takes a significant amount of time then we don't need an immediate start of next routine. Also, I don't think this is a new pattern in the project source. See `dbpurge.go`.

Reply:
https://pkg.go.dev/time#Ticker

> The ticker will adjust the time interval or drop ticks to make up for slow receivers.

I believe that will account for this situation, no?

Reply:
Interesting, do you know how ticker would know how much time the routine spent doing some logic?

Reply:
I think it depends if we want the reports to run every interval, or if we want at least interval between executions of report.

https://go.dev/play/p/ZOQOKPIWD58

```
2009-11-10 23:00:01 +0000 UTC m=+1.000000001
1.5s <-- thing take 500ms
2009-11-10 23:00:02.5 +0000 UTC m=+2.500000001
2s   <-- thing take 1000ms
2009-11-10 23:00:04.5 +0000 UTC m=+4.500000001
2.5s <-- thing takes 1500ms
```

If you comment out the `tick.Reset()` then you will see:

```
2009-11-10 23:00:01 +0000 UTC m=+1.000000001
1.5s <- thing take 500ms
2009-11-10 23:00:02 +0000 UTC m=+2.000000001
1.5s <- thing takes 1000ms
2009-11-10 23:00:03 +0000 UTC m=+3.000000001
1.5s <- thing takes 1500ms
2009-11-10 23:00:04 +0000 UTC m=+4.000000001
2s   <-- thing takes 2000ms, ticker slows down
```

Reply:
I'm going to leave it as is, but applied the suggestion from Mathias:

```go
	ticker := clk.NewTicker(delay)
	ticker.Stop()
```